### PR TITLE
Add MOC Filter

### PR DESCRIPTION
**Pull request by smcguire-cmu**
Adds the `MOCSearch` class, and the `search_moc` method to the catalog that allows filtering a catalog by a mocpy object.

Also adds a default `search` implementation to the `HealpixDataset` class. Previously searching was only possible on the catalog subclass.

### CODE REVIEW

File: src/lsdb/catalog/dataset/healpix_dataset.py [hunk 1]
```diff
@@ -194,6 +194,22 @@ def _perform_search(
         ddf_partition_map = {pixel: i for i, pixel in enumerate(filtered_pixels)}
         return ddf_partition_map, filtered_partitions_ddf
 
+    def search(self, search: AbstractSearch):
```
Comment (camposandro) [inline src/lsdb/catalog/dataset/healpix_dataset.py:197]: Can we delete the `search` method from the Catalog object entirely and just use this one to reduce duplication? We'd just need to be careful about margin filtering.

Comment (camposandro) [reply src/lsdb/catalog/dataset/healpix_dataset.py:197]: We could also remove it from `MarginCatalog`:

https://github.com/astronomy-commons/lsdb/blob/34f7d9d518533f52e25f1d21d5db02eec26fefde/src/lsdb/catalog/margin_catalog.py#L28

File: src/lsdb/core/search/moc_search.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+if TYPE_CHECKING:
+    from lsdb.types import HCCatalogTypeVar
+
+
+class MOCSearch(AbstractSearch):
+    """Filter the catalog by HEALPix pixels.
+
+    Filters partitions in the catalog to those that are in a specified pixel set.
+    Does not filter points inside those partitions.
```
Comment (camposandro) [inline src/lsdb/core/search/moc_search.py:20]: Filter the catalog by a MOC

File: src/lsdb/catalog/dataset/healpix_dataset.py (hunk 1 above)
Comment (smcguire-cmu) [reply src/lsdb/catalog/dataset/healpix_dataset.py:197]: The catalog has the margin logic, but it should call the superclass, but yeah removing the margin is a good idea

**Review by camposandro — APPROVED** [review]

**Pull request merged**